### ISSUE TITLE
Unable to modify list based beacons in salt:2017.7 branch

### ISSUE THREAD

**Issue by golmaal** [saltstack/salt#45312]
### Description of Issue/Question
Beacons such as diskusage cannot be modified. This issue requests back-porting relevant parts of @garethgreenaway beacons' re-write in develop to 2017.7. Please see below for a very specific change which addresses the issue.

### Setup
Setup salt-master and salt-minion on a docker using bootstrap-salt.sh git 2017.7 method.
Install python-psutil

### Steps to Reproduce Issue
[root@2017 salt]# salt salt beacons.list
```
salt:
    ----------
    beacons:
        ----------
```
[root@2017 salt]# salt salt beacons.add diskusage "[{'/var':50%}, {'interval':60}]"
```
salt:
    ----------
    comment:
        Added beacon: diskusage.
    result:
        True
```
[root@2017 salt]# salt salt beacons.list
```
salt:
    beacons:
      diskusage:
      - /var: 50%
      - interval: 60
```
[root@2017 salt]# salt salt beacons.save
```
salt:
    ----------
    comment:
        Beacons saved to /etc/salt/minion.d/beacons.conf.
    result:
        True
```
[root@2017 salt]# cat /etc/salt/minion.d/beacons.conf
```
beacons:
  diskusage:
  - /var: 50%
  - interval: 60
```
Trying to add /tmp to the diskusage beacon fails - 
[root@2017 salt]# salt salt beacons.modify diskusage "[{'/var':50%}, {'/tmp':50%}, {'interval':60}]"
```
salt:
    The minion function caused an exception: Traceback (most recent call last):
      File "/usr/lib/python2.7/site-packages/salt/minion.py", line 1493, in _thread_return
        return_data = executor.execute()
      File "/usr/lib/python2.7/site-packages/salt/executors/direct_call.py", line 28, in execute
        return self.func(*self.args, **self.kwargs)
      File "/usr/lib/python2.7/site-packages/salt/modules/beacons.py", line 261, in modify
        for (key, value) in sorted(_current.items())]
    AttributeError: 'list' object has no attribute 'items'
```

This happens because 2017.7 branch is missing the six.itemitems fix by @garethgreenaway in develop line. If I replace lines https://github.com/saltstack/salt/blob/2017.7/salt/modules/beacons.py#L260 to https://github.com/saltstack/salt/blob/2017.7/salt/modules/beacons.py#L267 by the following (from develop branch) the issue is resolved. I could open a pull request, but would like to check with @garethgreenaway  if anything else should be back-ported from his re-write of beacons in develop branch. 

**Sorry the following works with ps and diskusage beacons and breaks inotify beacons :(**
```
        _current_lines = []
        for _item in _current:
            _current_lines.extend(['{0}:{1}\n'.format(key, value)
                                  for (key, value) in six.iteritems(_item)])
        _new_lines = []
        for _item in _new:
            _new_lines.extend(['{0}:{1}\n'.format(key, value)
                              for (key, value) in six.iteritems(_item)])
        _diff = difflib.unified_diff(_current_lines, _new_lines)

```
After applying the fix above - 
[root@2017 salt]# salt salt beacons.modify diskusage "[{'/var':50%}, {'/tmp': 50%}, {'interval':60}]"
```
salt:
    ----------
    changes:
        ----------
        diff:
            ---
            +++
            @@ -1,2 +1,3 @@
             /var:50%
            +/tmp:50%
             interval:60
    comment:
        Modified beacon: diskusage.
    result:
        True
```
### Versions Report
[root@2017 salt]# salt --versions-report
Salt Version:
           Salt: 2017.7.0-n/a-dab4a8c

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.5 (default, Aug  4 2017, 00:39:18)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: centos 7.4.1708 Core
         locale: ANSI_X3.4-1968
        machine: x86_64
        release: 3.10.0-693.11.6.el7.x86_64
         system: Linux
        version: CentOS Linux 7.4.1708 Core

**Comment by golmaal** [comment]
Hi Gareth, the solution I tried to use worked with ps and diskusage, but broke for inotify. I guess there more to this problem then I originally though. Please try inotify and diskusage beacons for your solution in 2017.7 branch.

**Comment by garethgreenaway** [comment]
@golmaal Thanks for the report. Unfortunately the fix for this isn't as simple updating that one line to use six.iteritems instead pf ones.  The problem is related to the fact that prior to the upcoming Oxygen release we had beacons whose configuration that could be dictionary based or list based, this obviously causes issues when trying to use the beacon module to manage modules.  Things work as expected in the `oxygen` branch and `develop` because of https://github.com/saltstack/salt/pull/42317.

@saltstack/team-core Thoughts on back-porting PR #42317 to 2017.7?

**Comment by rallytime** [comment]
I'd rather not back-port such a large change/refactor into a point-release branch.

**Comment by garethgreenaway** [comment]
I'm leaning in the same direction it does leave the beacon module broken in certain situations for certain beacons.

**Comment by golmaal** [comment]
We're targeting to use 2017.7.3 assuming it's going to be released soon. I've written wrappers to delete followed by an add to simulate a consistent modify behavior. Thanks for your help Gareth.

**Comment by arizvisa** [comment]
Hey, was just surfing for issues with beacons and saw a mention of inotify here. I included some comments in issue #48698 which mentions an issue with the inotify validate function in 2019.2. Another thing that is mentioned is that none of the beacons are actually processed in the minion at startup due to a seemingly undocumented option "beacons_before_connect".

I'd paste a sumamry here, but it's kind of large.

**Comment by garethgreenaway** [comment]
@arizvisa Thanks.  Will take a look at the new comments in that issue.
